galsim/compound.py

```diff
@@ -645,6 +744,33 @@ def __str__(self):
         s += ')'
         return s
 
+    def _prepareDraw(self):
+        self._orig_obj._prepareDraw()
+        self.SBProfile = galsim._galsim.SBAutoConvolve(self._orig_obj.SBProfile, self._real_space,
+                                                       self._gsparams)
+
+    def shoot(self, n_photons, rng=None):
+        """Shoot photons into a PhotonArray.
+
+        @param n_photons    The number of photons to use for photon shooting.
+        @param rng          If provided, a random number generator to use for photon shooting,
+                            which may be any kind of BaseDeviate object.  If `rng` is None, one
+                            will be automatically created, using the time as a seed.
+                            [default: None]
+        @returns PhotonArray.
+        """
+        # Setup the rng if not provided one.
+        if rng is None:
```

Comment:
I've been having some fun trying this out.  My thought is to use `bp::make_constructor` and essentially do runtime type introspection in the pointed to construct function. E.g., for BaseDeviate the construct function and wrapping would look like:

```c++
        static boost::shared_ptr<BaseDeviate> construct(const bp::object &seed) {
            if (seed.ptr() == Py_None) {
                return boost::shared_ptr<BaseDeviate>(new BaseDeviate(0l));
            }
            bp::extract<long> xl(seed);
            if (xl.check()) {
                return boost::shared_ptr<BaseDeviate>(new BaseDeviate(xl()));
            }
            bp::extract<const BaseDeviate&> xbd(seed);
            if (xbd.check()) {
                return boost::shared_ptr<BaseDeviate>(new BaseDeviate(xbd()));
            }
            bp::extract<std::string> xs(seed);
            if (xs.check()) {
                return boost::shared_ptr<BaseDeviate>(new BaseDeviate(xs()));
            }
            throw std::runtime_error("Cannot construct BaseDeviate from given seed.");
        }

        static void wrap() {
            bp::class_<BaseDeviateCallBack>
                pyBaseDeviate("BaseDeviate", "", bp::no_init);
            pyBaseDeviate
                // .def(bp::init<long>(bp::arg("seed")=0))
                // .def(bp::init<const BaseDeviate&>(bp::arg("seed")))
                // .def(bp::init<std::string>(bp::arg("seed")))
                .def("__init__",
                    bp::make_constructor(
                        &construct, bp::default_call_policies(),
                        (bp::arg("seed")=bp::object())
                    ))
                ...
```

This mostly works, except for the virtual override of `_val` in `DistDeviate`, which I'm guessing is the tricky bit you were eluding to above.  I found [this stackoverflow answer](http://stackoverflow.com/questions/19254715/virtual-override-in-boost-python-objects-created-by-static-method-or-factory) which looks relevant, but I thought I'd ask if I'm missing something simpler before tackling that.